### ISSUE TITLE
Bug: [@typescript-eslint/no-unnecessary-type-assertion] False positives when using non-null assertion operator for functions with potentially void return type

### ISSUE THREAD

**Issue by lukaVarga** [typescript-eslint/typescript-eslint#8906]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=GYVwdgxgLglg9mABBOAnVBTaAbAngEQyi1jAHMA5DDAE1oEEBnRjVWBAdRigAsBVMHWAwwtAEpEQqMADUAhtgAUc1GQD8ALkRgQAWwBGrAJRadB1ogA%2BicEJG1EAbwBQiRDGCJlqo09dvETCgpJBUyAG5-AF9nGOdQSHYkFHQSPEJiaBFKajoaJhY2eDAuXgoQbGwJYOl5JTDNbT1DVBMm81QrJsq-Nw8vMN8XAMDJaUQwyLc4tyCQ7uxIuISshHdINEwcAiISbKpaBmZWJNKeGTgYGmqQuu91U2bjR46ugDdLml73T3uh-1mY1CqimiBicRQYEYUEQxGhAGUQBAIBhmC8WogALzITZpHaZUg5Q75Y5FTjcfiCDDCUTXIF3IwAQkikOhsNRUERyNRjAATOiLNiUlsoOldllyAc8gUTsUzuVKjdagpFEzIgB6dWIbAiYidRg8OAVGhgADkMOAchg2EQPFYGGcrJhcKgAFF0GgBZ1sSJhXiMntJbkjoVThSLlclbIVWrnJrEAAVeEAYS8fu2vkt1sYjoQbJdSeT7tQnvaGJ9G1S2wDEqJ0tJYd4EbpNWjSiMkWc2CIiDeKgTHIAjF73p9It2YX3UAPofyyxZrDpKuOe1OZ1AAMwj6y2an2GidtdDrHrdOi-GBush2XkpufKMMyJH2cns9ign7YMk0NyikKqr0jGT79hyG6vri1bioSUrXmSJQUgIdi0g%2BQHOPE4CrEgmAAI4gDAmA0OUHSKAA%2BowcC6BgRFem0HxXE4YJoTheEEURLSKM%2BUCDoyHZxlqBpGtgXxZtgzjMfhtBsawHEgdCg68WJGC4RJhFPKgMnThyvI8Rq-GGsaiAiYpymsWpGnrryCniaZxGcRuOl8YgAkGUZ1mSWZdkdkAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function incorrectlyDetectingNeededAssertionWithVoidReturnVal(arg?: number): number | void {
  if (arg) {
    return arg;
  }
}

const testError: number = incorrectlyDetectingNeededAssertionWithVoidReturnVal()!;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The rule should correctly detect the type assertion as necessary since the function may return `void` and we are trying to assign the result of the function to a constant that is a `number`.

### Actual Result

Linter fails as it detects the type assertion as unnecessary. By omitting the non-null operator, TSC (correctly) reports an error (which you can see in a line below)

### Additional Info

The rule works as expected if the function potentially returns `null` or `undefined` - it only reports false positives if the function returns `void`. It also reports false positives when using class methods that potentially return void.

The rule also works correctly when checking the usage of non-null operator with variables / constants - it does not matter whether they are potentially `null`, `undefined` or `void`.
